File: polyglot/go/binary_parser.go
package main

import (
	"archive/zip"
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
)

// PEHeader represents the complete PE header structure for parsing
type PEHeader struct {
	DOSHeader  DOSHeader
	PESignature uint32
	PEHeader   PEStandardHeader
	Optional    PEOptionalHeader
	Sections    []SectionInfo
}

// DOSHeader is the MZ header at the start of every PE file
type DOSHeader struct {
	Magic     [2]byte
	Offset    uint16 // Offset to PE header (eip)
	Subsystem uint16 // Subsystem type
}

// PEStandardHeader contains the standard PE fields
type PEStandardHeader struct {
	Signature   uint32
	Machine      uint16
	NumberOfSections uint16
	TimeDateStamp  uint32
	PointerToSymbolTable  uint32
	NumberOfSymbols       uint32
	OptionalHeaderSize    uint16
	Characteristics       uint16
}

// PEOptionalHeader contains the optional header fields
type PEOptionalHeader struct {
	Magic              uint16
	MajorLinkerVersion byte
	MinorLinkerVersion byte
	SizeOfCode         uint32
	SizeOfInitializedData  uint32
	SizeOfUninitializedData   uint32
	AddressOfEntryPoint    uint32
	BaseOfCode            uint32
	BaseOfData            uint32
	ImageBase             uint32
	SectionAlignment      uint32
	FileAlignment         uint32
	MajorOSVersion        uint16
	MinorOSVersion        uint16
	MajorImageVersion     uint16
	MinorImageVersion     uint16
	MajorSubsystemVersion  uint16
	MinorSubsystemVersion  uint16
}

// SectionInfo represents a PE section
type SectionInfo struct {
	Name       string
	VirtualSize   uint32
	VirtualAddress uint32
	SizeOfRawData    uint32
	PointerToRawData uint32
	Characteristics  uint32
}

// PackerResult is the main result structure containing all analysis data
type PackerResult struct {
	Path           string
	FileName       string
	FileSize       int64
	Magic          [2]byte
	DOSOffset      uint16
	PESignature    uint32
	Machine        uint16
	SectionsCount  uint16
	OptionalMagic   uint16
	Entropy        float64
	ShannonEntropy float64
	IsPacked       bool
	PackersDetected []string
	Resources      ResourceInfo
	Sections       []SectionInfo
}

// ResourceInfo contains UPX resource information if found
type ResourceInfo struct {
	HasUPXResource   bool
	UPXVersion       string
	UPXOffset        uint32
	UPXSize          uint32
	UPXCompression   uint16
}

// Constants for PE headers
const (
	PESignature = 0x4550 // "PE\0\0"
	MZSignature  = 0x5A4D // "MZ"
	IMAGE_NT_OPTIONAL_HDR_MAGIC_PE32     = 0x10b
	IMAGE_NT_OPTIONAL_HDR_MAGIC_PE32PLUS = 0x20b
)

// Constants for UPX resources
const (
	UPX_RESOURCE_DIRECTORY_SIZE = 8
	UPX_VERSION_4_0             = 0x0400
)

// CalculateShannonEntropy computes the Shannon entropy of a byte slice
func CalculateShannonEntropy(data []byte) float64 {
	if len(data) == 0 {
		return 0.0
	}

	freq := make(map[byte]int, 256)
	for _, b := range data {
		freq[b]++
	}

	var entropy float64 = 0.0
	total := float64(len(data))

	for count := range freq {
		if count > 0 {
			p := float64(count) / total
			entropy -= p * math.Log2(p)
		}
	}

	return entropy
}

// ReadPEHeader reads and parses the PE header from a file
func ReadPEHeader(filePath string) (*PEHeader, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("open file: %w", err)
	}
	defer f.Close()

	data, err := io.ReadAll(f)
	if err != nil {
		return nil, fmt.Errorf("read file: %w", err)
	}

	header := &PEHeader{
		DOSHeader: DOSHeader{
			Magic:  [2]byte{'Z', 'M'}, // Little-endian MZ
		},
		PESignature: PESignature,
	}

	if len(data) < 64 {
		return nil, fmt.Errorf("file too small for PE header")
	}

	// Parse DOS header (little-endian)
	binary.Read(bytes.NewReader(data[0:2]), binary.LittleEndian, &header.DOSHeader.Magic)
	header.DOSOffset = binary.LittleEndian.Uint16(data[48:50])
	header.PESignature = binary.LittleEndian.Uint32(data[60:64])

	// Parse PE standard header
	if len(data) < 64+header.DOSOffset {
		return nil, fmt.Errorf("PE header offset out of bounds")
	}

	binary.Read(bytes.NewReader(data[header.DOSOffset:]), binary.LittleEndian, &header.PESignature)
	binary.Read(bytes.NewReader(data[header.DOSOffset+2:]), binary.LittleEndian, &header.PEHeader)

	// Parse optional header if present
	if len(data) >= int(header.DOSOffset)+int(header.PEHeader.OptionalHeaderSize) {
		optStart := header.DOSOffset + 40 // PE standard header is 40 bytes
		binary.Read(bytes.NewReader(data[optStart:]), binary.LittleEndian, &header.Optional.Magic)
	}

	// Enumerate sections
	header.Sections = EnumerateSections(data, header.PEHeader, header.Optional)

	return header, nil
}

// EnumerateSections extracts all section information from the PE file
func EnumerateSections(data []byte, stdHeader PEStandardHeader, optHeader PEOptionalHeader) []SectionInfo {
	var sections []SectionInfo

	if len(data) < int(stdHeader.PointerToSymbolTable)+int(optHeader.SizeOfCode) {
		return sections
	}

	// Calculate section table offset
	sectionOffset := 64 + stdHeader.OptionalHeaderSize // After DOS + PE standard header

	if len(data) <= int(sectionOffset) {
		return sections
	}

	for i := uint16(0); i < stdHeader.NumberOfSections; i++ {
		nameStart := sectionOffset + (i * 40)
		
		var name [8]byte
		binary.Read(bytes.NewReader(data[nameStart:nameStart+4]), binary.LittleEndian, &name[:4])
		nameStr := string(name[:4])

		if len(nameStr) < 2 {
			continue
		}

		// Pad to 8 bytes for proper display
		for j := 4; j < 8; j++ {
			if name[j] == 0 {
				name[j] = ' '
			}
		}

		sections = append(sections, SectionInfo{
			Name:       strings.TrimSpace(string(name[:])),
			VirtualSize: binary.LittleEndian.Uint32(data[nameStart+4 : nameStart+8]),
			VirtualAddress: binary.LittleEndian.Uint32(data[nameStart+8 : nameStart+12]),
			SizeOfRawData:  binary.LittleEndian.Uint32(data[nameStart+16 : nameStart+20]),
			PointerToRawData: binary.LittleEndian.Uint32(data[nameStart+20 : nameStart+24]),
			Characteristics: binary.LittleEndian.Uint32(data[nameStart+28 : nameStart+32]),
		})
	}

	return sections
}

// DetectUPXResources scans for UPX resource directory
func DetectUPXResources(data []byte) ResourceInfo {
	var res ResourceInfo

	if len(data) < 64 {
		return res
	}

	// Check if file is large enough to have resources
	if len(data) < 256 {
		return res
	}

	// Look for UPX resource directory at offset 0x18 (24 bytes into PE header)
	resourceOffset := 64 + 24 // DOS(64) + PE Standard(24)

	if len(data) < int(resourceOffset)+UPX_RESOURCE_DIRECTORY_SIZE {
		return res
	}

	// Read resource directory header
	resDirStart := resourceOffset
	binary.Read(bytes.NewReader(data[resDirStart:]), binary.LittleEndian, &res.UPXVersion)

	// UPX 4.0 has specific signature at offset 0x18
	if len(data) >= int(resourceOffset)+UPX_RESOURCE_DIRECTORY_SIZE {
		version := binary.LittleEndian.Uint32(data[resourceOffset : resourceOffset+4])
		
		if version == UPX_VERSION_4_0 || version == UPX_VERSION_4_1 {
			res.HasUPXResource = true
			res.UPXVersion = fmt.Sprintf("v%d.%d", byte(version>>8), byte(version&0xFF))
			
			// Extract UPX metadata from resource directory
			res.UPXOffset = binary.LittleEndian.Uint32(data[resourceOffset+4 : resourceOffset+8])
			res.UPXSize = binary.LittleEndian.Uint32(data[resourceOffset+8 : resourceOffset+12])
			res.UPXCompression = binary.LittleEndian.Uint16(data[resourceOffset+12 : resourceOffset+14])
		}
	}

	return res
}

// ScanForPackerSignatures looks for known packer signatures in the file
func ScanForPackerSignatures(data []byte, header *PEHeader) []string {
	var detected []string

	signatures := map[string][]byte{
		"UPX":       {"UPX0", "UPX1"},
		"ASPack":    {"ASPak", "ASPak2"},
		"Themida":   {"ThmD", "ThmD2"},
		"VMProtect": {"VMP_", "VMPr"},
		"MPRESS":    {"MPre", "MPreS"},
	}

	for name, patterns := range signatures {
		for _, pattern := range patterns {
			if len(data) >= 8 && bytes.Contains(data[:8], pattern) {
				detected = append(detected, fmt.Sprintf("%s (signature match)", name))
			}
		}
	}

	return detected
}

// ParseBinaryFile is the main entry point for binary analysis
func ParseBinaryFile(filePath string) (*PackerResult, error) {
	result := &PackerResult{
		Path:     filePath,
		Magic:    [2]byte{'Z', 'M'},
		DOSOffset: 64, // Default MZ offset
		PESignature: PESignature,
	}

	f, err := os.Open(filePath)
	if err != nil {
		return result, fmt.Errorf("open file: %w", err)
	}
	defer f.Close()

	data, err := io.ReadAll(f)
	if err != nil {
		return result, fmt.Errorf("read file: %w", err)
	}

	result.FileSize = int64(len(data))
	result.FileName = filepath.Base(filePath)

	// Parse PE header if present
	if len(data) >= 64 && bytes.Equal(data[0:2], []byte{'Z', 'M'}) {
		header, err := ReadPEHeader(filePath)
		if err == nil {
			result.DOSOffset = header.DOSOffset
			result.PESignature = header.PESignature
			result.Machine = header.PEHeader.Machine
			result.SectionsCount = header.PEHeader.NumberOfSections
			result.OptionalMagic = header.Optional.Magic
			result.Sections = header.Sections

			// Detect packers via signatures
			signatures := ScanForPackerSignatures(data, header)
			if len(signatures) > 0 {
				result.IsPacked = true
				result.PackersDetected = append(result.PackersDetected, signatures...)
			}

			// Check for UPX resources
			resources := DetectUPXResources(data)
			result.Resources = resources
			if resources.HasUPXResource {
				result.IsPacked = true
				result.PackersDetected = append(result.PackersDetected, 
					fmt.Sprintf("UPX %s (resources: offset=%d, size=%d)",
						resources.UPXVersion, resources.UPXOffset, resources.UPXSize))
			}

			// Calculate entropy
			result.ShannonEntropy = CalculateShannonEntropy(data)
			
			// High entropy often indicates packing/compression
			if result.ShannonEntropy > 7.5 {
				result.IsPacked = true
				result.PackersDetected = append(result.PackersDetected, 
					fmt.Sprintf("High entropy: %.2f (possible compression)", result.ShannonEntropy))
			}
		}
	}

	return result, nil
}

// OutputResult formats the PackerResult as JSON
func OutputResult(result *PackerResult) {
	jsonBytes, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "JSON marshal error: %v\n", err)
		return
	}

	fmt.Println(string(jsonBytes))
}

func main() {
	// Demo with a sample binary file
	samplePath := "./test_binary.exe"

	if _, err := os.Stat(samplePath); err == nil {
		result, err := ParseBinaryFile(samplePath)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error parsing: %v\n", err)
			os.Exit(1)
		}

		OutputResult(result)
	} else {
		// Create a minimal PE file for testing if sample doesn't exist
		fmt.Println("Sample binary not found. Creating test file...")
		
		testPE := createMinimalPE()
		err := os.WriteFile("./test_binary.exe", testPE, 0644)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error creating test file: %v\n", err)
			os.Exit(1)
		}

		result, err := ParseBinaryFile("./test_binary.exe")
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error parsing test: %v\n", err)
			os.Exit(1)
		}

		OutputResult(result)
	}
}

// createMinimalPE creates a minimal valid PE file for testing
func createMinimalPE() []byte {
	var buf bytes.Buffer

	// DOS Header (64 bytes)
	buf.WriteString("MZ") // Magic
	binary.Write(&buf, binary.LittleEndian, uint16(0x8000)) // eip offset
	binary.Write(&buf, binary.LittleEndian, uint16(2))     // subsystem

	// PE Header (56 bytes)
	binary.Write(&buf, binary.LittleEndian, uint32(PESignature))
	binary.Write(&buf, binary.LittleEndian, uint16(0x14c))  // machine (i386)
	binary.Write(&buf, binary.LittleEndian, uint16(1))     // sections
	binary.Write(&buf, binary.LittleEndian, uint32(0x12345678)) // timestamp
	binary.Write(&buf, binary.LittleEndian, uint32(0))      // symbol table
	binary.Write(&buf, binary.LittleEndian, uint16(0))      // symbols
	binary.Write(&buf, binary.LittleEndian, uint16(40))     // optional header size
	binary.Write(&buf, binary.LittleEndian, uint16(0x2202)) // characteristics

	// Optional Header (28 bytes for PE32)

File: polyglot/csharp/binary_parser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Diagnostics;

namespace PackPeek.Core.Parsers
{
    /// <summary>
    /// Complete binary parser for detecting packers (UPX, ASPack, Themida, MPRESS, VMProtect)
    /// and calculating section entropy. Emits YARA and SARIF output formats.
    /// </summary>
    public class BinaryParser
    {
        private readonly byte[] _data;
        private readonly string _filename;

        // Known magic signatures for packers (offset 0x40 is common for UPX)
        private static readonly Dictionary<string, Func<byte[], int>> PackerSignatures = new()
        {
            ["UPX"] = (data) => FindSignature(data, "UPX!"),
            ["ASPack"] = (data) => FindString(data, "ASPack"),
            ["Themida"] = (data) => FindString(data, "Themida") || FindString(data, "THEMIDA"),
            ["MPRESS"] = (data) => FindSignature(data, 0x4D505245), // "MPRE"
            ["VMProtect"] = (data) => FindSignature(data, 0x564D5052), // "VMPR"
        };

        public BinaryParser(byte[] data, string filename = null)
        {
            _data = data;
            _filename = filename ?? Path.GetFileName("unknown");
        }

        /// <summary>
        /// Main entry point for parsing a binary file.
        /// </summary>
        public ParsedResult Parse()
        {
            var result = new ParsedResult(_filename);

            // 1. Basic header analysis
            AnalyzeHeader(result);

            // 2. Section entropy calculation
            CalculateSectionEntropy(result);

            // 3. Packer detection
            DetectPackers(result);

            // 4. String extraction for additional clues
            ExtractSuspiciousStrings(result);

            return result;
        }

        private void AnalyzeHeader(ParsedResult result)
        {
            if (_data.Length == 0)
                return;

            // Check file size anomalies (packed files often have specific sizes)
            var fileSize = _data.Length;
            result.FileSize = fileSize;

            // Calculate overall entropy as a quick indicator
            result.OverallEntropy = CalculateShannonEntropy(_data);
        }

        private void CalculateSectionEntropy(ParsedResult result)
        {
            if (_data.Length == 0)
                return;

            // Parse PE headers to get sections (simplified for demo - assumes valid PE)
            var peHeaderOffset = 64; // Standard PE header offset after DOS header
            if (peHeaderOffset >= _data.Length || !_IsPE(_data))
                return;

            // Read PE signature
            string peSig = Encoding.ASCII.GetString(_data, peHeaderOffset, 4);
            if (!peSig.Equals("PE\x00\x00", StringComparison.Ordinal))
                return;

            // Parse Optional Header to get section table
            int optionalHeaderStart = peHeaderOffset + 24; // After PE signature and fields
            ushort machineType = BitConverter.ToUInt16(_data, optionalHeaderStart);
            
            // Skip to Section Table RVA (offset 0x38 in optional header)
            int sectionTableRVA = BitConverter.ToInt32(_data, optionalHeaderStart + 0x38);
            
            if (sectionTableRVA == 0 || sectionTableRVA >= _data.Length)
                return;

            // Calculate file offset for section table: PE header + fields before sections
            int sectionTableOffset = peHeaderOffset + 64 + BitConverter.ToInt16(_data, optionalHeaderStart + 0x3C);
            
            if (sectionTableOffset >= _data.Length)
                return;

            // Read section count and names
            ushort sectionCount = BitConverter.ToUInt16(_data, sectionTableOffset);
            var sections = new List<SectionInfo>();

            for (int i = 0; i < sectionCount && i * 40 < _data.Length; i++)
            {
                int offset = sectionTableOffset + (i * 40);
                string name = Encoding.ASCII.GetString(_data, offset, 8).TrimEnd('\0');
                uint virtualSize = BitConverter.ToUInt32(_data, offset + 16);
                uint rawSize = BitConverter.ToUInt32(_data, offset + 20);

                sections.Add(new SectionInfo
                {
                    Name = name,
                    VirtualAddress = BitConverter.ToInt32(_data, optionalHeaderStart + 124), // Simplified
                    Size = (int)rawSize,
                });
            }

            foreach (var section in sections)
            {
                if (section.Size > 0 && section.Size < _data.Length)
                {
                    var offset = section.VirtualAddress;
                    var slice = new byte[section.Size];

                    // Handle virtual address vs file offset
                    if (offset >= _data.Length)
                        continue;

                    int end = Math.Min(offset + section.Size, _data.Length);
                    Array.Copy(_data, offset, slice, 0, end - offset);

                    var entropy = CalculateShannonEntropy(slice);
                    result.Sections.Add(new SectionEntropyResult
                    {
                        Name = section.Name,
                        VirtualSize = section.Size,
                        Entropy = entropy,
                        HighEntropyThreshold = 7.5, // Typical for packed sections
                    });
                }
            }
        }

        private void DetectPackers(ParsedResult result)
        {
            var findings = new List<PackerFinding>();

            foreach (var kvp in PackerSignatures)
            {
                try
                {
                    int offset = kvp.Value(_data);
                    if (offset >= 0 && offset < _data.Length)
                    {
                        findings.Add(new PackerFinding
                        {
                            Name = kvp.Key,
                            Type = "MagicHeader",
                            Offset = offset,
                            Confidence = 0.95m,
                        });
                    }
                }
                catch (Exception ex)
                {
                    // Silently handle parse errors
                }
            }

            result.Packers = findings;
        }

        private void ExtractSuspiciousStrings(ParsedResult result)
        {
            var suspiciousPatterns = new[]
            {
                "UPX", "ASPack", "Themida", "MPRESS", "VMProtect",
                ".text", ".rdata", ".idata", ".reloc",
                "DllMain", "WinMain", "CRT"
            };

            foreach (var pattern in suspiciousPatterns)
            {
                try
                {
                    int offset = FindString(_data, pattern);
                    if (offset >= 0)
                    {
                        result.Strings.Add(new StringFinding
                        {
                            Pattern = pattern,
                            Offset = offset,
                            Type = "Suspicious",
                        });
                    }
                }
                catch
                {
                    // Ignore errors
                }
            }
        }

        private static int CalculateShannonEntropy(byte[] data)
        {
            if (data == null || data.Length == 0)
                return 0;

            var counts = new byte[256];
            foreach (var b in data)
                counts[b]++;

            double entropy = 0.0;
            int total = data.Length;

            for (int i = 0; i < 256; i++)
            {
                if (counts[i] > 0)
                {
                    double p = (double)counts[i] / total;
                    entropy -= p * Math.Log(p, 2);
                }
            }

            return (int)(entropy * 100.0); // Return as integer percentage for easier comparison
        }

        private static int FindSignature(byte[] data, string signature)
        {
            var sigBytes = Encoding.ASCII.GetBytes(signature);
            if (sigBytes.Length > data.Length)
                return -1;

            for (int i = 0; i <= data.Length - sigBytes.Length; i++)
            {
                bool match = true;
                for (int j = 0; j < sigBytes.Length; j++)
                {
                    if (data[i + j] != sigBytes[j])
                    {
                        match = false;
                        break;
                    }
                }

                if (match)
                    return i;
            }

            return -1;
        }

        private static int FindString(byte[] data, string pattern)
        {
            var sigBytes = Encoding.ASCII.GetBytes(pattern);
            if (sigBytes.Length > data.Length)
                return -1;

            for (int i = 0; i <= data.Length - sigBytes.Length; i++)
            {
                bool match = true;
                for (int j = 0; j < sigBytes.Length; j++)
                {
                    if (data[i + j] != sigBytes[j])
                    {
                        match = false;
                        break;
                    }
                }

                if (match)
                    return i;
            }

            return -1;
        }

        private static bool _IsPE(byte[] data)
        {
            if (data.Length < 64)
                return false;

            string peSig = Encoding.ASCII.GetString(data, 64, 4);
            return peSig.Equals("PE\x00\x00", StringComparison.Ordinal);
        }

        /// <summary>
        /// Generates YARA rules from parser findings.
        /// </summary>
        public static string GenerateYaraRules(ParsedResult result)
        {
            var sb = new StringBuilder();
            sb.AppendLine("// Auto-generated YARA rules for: " + result.FileName);
            sb.AppendLine("#pragma once");
            sb.AppendLine("");

            // UPX rule
            if (result.Packers.Any(p => p.Name == "UPX"))
            {
                sb.AppendLine("rule upx_detected");
                sb.AppendLine("{");
                sb.AppendLine("\tmeta:");
                sb.AppendLine("\t\tdescription = \"UPX packer detected\";");
                sb.AppendLine("\t\tauthor = \"PackPeek\";");
                sb.AppendLine("\t\tdate = \"" + DateTime.Now.ToString("yyyy-MM-dd") + "\";");
                sb.AppendLine("\t\ttags = {upx